### PR TITLE
[gpl]: Improve performance of GPL by 13%

### PR DESCRIPTION
**Pull request by QuantamHD**
Removed binStor concept. Having a vector of Bin objects is faster due to cache locality considerations.

### CODE REVIEW

**Comment by antonblanchard** [comment]
Nice speed up @QuantamHD!

File: src/gpl/src/nesterovBase.h [hunk 1]
```diff
@@ -737,13 +707,13 @@ class BinGrid
   std::pair<int, int> getMinMaxIdxX(const Instance* inst) const;
   std::pair<int, int> getMinMaxIdxY(const Instance* inst) const;
 
-  const std::vector<Bin*>& bins() const;
+  std::vector<Bin>& bins();
+  const std::vector<Bin>& binsConst() const { return bins_; };
```
Comment (maliberty) [inline src/gpl/src/nesterovBase.h:711]: Why is the non-const version now needed?  (Similarly where this API propagates upwards)

File: src/gpl/src/nesterovBase.cpp [hunk 2]
```diff
@@ -773,19 +772,16 @@ void BinGrid::initBins()
 
   log_->info(GPL, 29, "BinSize: {} {}", binSizeX_, binSizeY_);
 
-  // initialize binStor_, bins_ vector
-  binStor_.resize(binCntX_ * binCntY_);
-  bins_.reserve(binCntX_ * binCntY_);
+  size_t bin_dimensions = static_cast<size_t>(binCntX_ * binCntY_);
```
[the remark this answers is not in the document]
Comment (maliberty) [reply src/gpl/src/nesterovBase.cpp:775]: An int is fine - if we have 2^32 bins we have bigger problems.

**Review by maliberty — CHANGES_REQUESTED** [review]
Surprisingly good gain!  Do you have a sense for how much was inlining vs data locality?

**Comment by maliberty** [comment]
Btw, on what design(s) was the 13% measured?

**Comment by QuantamHD** [comment]
medium06.tcl in gpl/test

File: src/gpl/src/nesterovBase.h (hunk 1 above)
Comment (QuantamHD) [reply src/gpl/src/nesterovBase.h:711]: bins is called in the non-const method `void NesterovBase::updateDensityForceBin()`

**Comment by QuantamHD** [comment]
@antonblanchard @maliberty I should not that I think this test is sans `-timing_driven -congestion_driven` which tend to dominate runtime.

**Comment by QuantamHD** [comment]
> Surprisingly

Mostly data locality. It was already inlining most of the lx, ly and friends.

File: src/gpl/src/nesterovBase.h (hunk 1 above)
Comment (maliberty) [reply src/gpl/src/nesterovBase.h:711]: I see the vector was constant but the pointers were not.  Now the vector itself must be mutable.

**Review by maliberty — APPROVED** [review]

**Pull request merged**